Cookie refactor same site v2 - replacing PR #1149

taniwallach opened the pull request:
Moved code to an updated branch of develop.

This is a resubmit of https://github.com/openwebwork/webwork2/pull/1149 to add secure cookie support and support for "same-site" as it was decided that additional fixes are needed.

Please see the discussion at  https://github.com/openwebwork/webwork2/pull/1149 .

---

Main points = refactor the Cookie code:

1. Use `CGI::Cookie` instead of `Apache2:Cookie`, as the new code needs support for the `samesite` attribute.
2. Added CGI::Cookie to bin/check_modules.pl and Dockerfile.
    - Note: The support for samesite dates to June 2019 in CGI::Cookie 4.45.
3. Remove obsolete, commented out, code using cookies from `lib/WeBWorK/ContentGenerator/Logout.pm`.
4. Drop the constant `COOKIE_LIFESPAN` and instead allow setting cookie lifespan using site / course environment configuration variables.
    - `$CookieLifeTime`  - for when cookie based session management **is** in use - default to 6 hours.
    - `$CookieLifeTime2` - for when cookie based session management is **not** in use, defaults to 30 days.
5. Allow setting value of cookies `samesite` and `secure` attribute using site / course environment configuration variable:
    - `$CookieSameSite`
    - `$CookieSecure`

taniwallach commented:
This still has the issue which was discoved by @drgrice1 after the code was modified to try to check the cookie did not expire... see https://github.com/openwebwork/webwork2/pull/1149#discussion_r593836887

I ran into problems on my first attempt to fix that, and will resume on Sunday.

drgrice1 commented on conf/defaults.config:
```diff
@@ -924,6 +924,35 @@ $session_management_via = "session_cookie";
+$CookieLifeTime = "+7d";
+
+# Set cookie lifetime for when $session_management_via ne "session_cookie"
+$CookieLifeTime2 = "+30d";
```
I was thinking that it would be better to just have one `$CookieLifeTime` variable here.  It is silly to have two variables that are used in exactly the same way for the same reason, but conditionally on the `session_management_via` setting.  Just document here that when using session management via keys you probably want a longer setting than when using session management via cookies.

taniwallach replied:
Done

taniwallach commented:
In https://github.com/openwebwork/webwork2/pull/1269/commits/dfba7fc7f3e4e743596f678032e6668badab5feb I reverted back to the old approach of determining whether they session key is still valid, which is used whether or not session management depends on the cookie or not.

The following commit has some additions to documentation inside the code/config files, and has some formatting changes.

drgrice1 commented on lib/WeBWorK/Authen.pm:
```diff
@@ -916,27 +883,37 @@ sub sendCookie {
+		$cookie->max_age( $lifetime2 );
+	} else {
+		$cookie->expires( $lifetime );
+		$cookie->max_age( $lifetime );
```
As I said in my other comment, I recommend getting rid of `CookieLifeTime2` (and `$lifetime2` here).  Furthermore, I recommend changing this to
```perl
if ($lifetime ne 'session') {
	$cookie->expires( $lifetime );
	$cookie->max_age( $lifetime );
}
```
Then add to the documentation to the config files explaining that the setting of `session` for `$CookieLifeTime` means that cookies will expire with the session.  Not calling `expires` and `max_age` is how you make that happen.

taniwallach replied:
Done

taniwallach replied:
Done

taniwallach commented:
@pstaabp - This is pretty similar to what was in https://github.com/openwebwork/webwork2/pull/1149 when you reviewed it (but with only one `$CookieLifeTime` setting).

Due to the bug explained there and below tPR https://github.com/openwebwork/webwork2/pull/1149 was reverted after it was merged.

The bug I introduced late on was in https://github.com/openwebwork/webwork2/pull/1149/commits/0280ff6780a884a4fdaa0269cb8340324a0a2ce when I modified  `lib/WeBWorK/Authen.pm` with
```
	if ( $ce->{session_management_via} eq "session_cookie"
		&& $ce->{CookieSecure} && defined($self->{cookie_timestamp}) ) {
		$timestampValid = ( time <= $self->{cookie_timestamp} + $ce->{CookieLifeTime} ); 

```
as explained in https://github.com/openwebwork/webwork2/pull/1149#discussion_r583057948 .

The problem is that the changes treated `$ce->{CookieLifeTime}` as if it was a numeric value in seconds, while it is a specially parsed string, and that triggers an error.

In any case, upon further thought - trusting the cookie's timestamp value which may be tampered with on the client side is not really a good idea. The potential of such abuse was mentioned in https://github.com/openwebwork/webwork2/pull/576#issuecomment-98900894 and that PR had added to code *not** to use the cookie's timestamp.

I'm pretty confident that this version is OK, but it needs someone to test it.

pstaabp reviewed (approved):
this is looking good.  I tested in Safari, which doesn't give an error, but does detect that SameSite is strict.  Also tested in FF and without this patch, lots of warnings, but with the patch FF seems happy.

drgrice1 commented on lib/WeBWorK/Authen.pm:
```diff
@@ -77,7 +76,7 @@ our $GENERIC_ERROR_MESSAGE = "";  # define in new
 #####################
 
-use constant COOKIE_LIFESPAN => 60*60*24*30; # 30 days
+# Replaced constant COOKIE_LIFESPAN with a course environment setting which defaults to 30 days.
```
We don't need comments referring to now non-existent constants.  So delete the comment.  It is also not correct for the actual defaults in defaults.config.  There the default is set to 7 days.

taniwallach replied:
Done.

drgrice1 commented on lib/WeBWorK/Cookie.pm:
```diff
@@ -35,13 +34,13 @@ use warnings;
-        require APR::Request::Error;
-		Apache2::Cookie->import;
-		push @WeBWorK::Cookie::ISA, "Apache2::Cookie";
+		#require APR::Request::Error;
```
This line is not needed anymore.  Delete!

drgrice1 commented on lib/WeBWorK/Authen.pm:
```diff
@@ -916,27 +883,40 @@ sub sendCookie {
-		-secure  => 0,
+
+	my $sameSite  = $ce->{CookieSameSite};
+	my $secure    = $ce->{CookieSecure};    # Warning: use 1 only if using https
```
There is no need to defined new variables here.  Just use `$ce->{CookieSameSite}` below in the one place that the variable `$sameSite` is used.  Same for `$secure`.

In general it is bad coding practice to define a variable if its value is only used once.  It is inefficient and clutters the variable namespace.  Furthermore, there is generally no need to define a variable for accessing a hash key.

You seem to do this sort of thing frequently, and the code could be made much more concise and efficient by not doing this.  It also improves readability of the code as it is easier to backtrack to where the value in use comes from.

taniwallach replied:
Done. I'll try to avoid doing this in the future.

drgrice1 commented on lib/WeBWorK/Authen.pm:
```diff
@@ -945,21 +925,25 @@ sub killCookie {
-		-secure => 0,
+
+	my $sameSite  = $ce->{CookieSameSite};
+	my $secure    = $ce->{CookieSecure};    # Warning: use 1 only if using https
```
Again, just use `$ce->{...}` below.

taniwallach replied:
Done

drgrice1 reviewed (approved):
This looks good now.  The previous authentication methods are preserved.  This basically just replaces the outdated Apache::Cookie module with CGI::Cookie, and adds the functionality that module provides.  Everything works as it should.

I added a few code clean up suggestions, but other than that this is looking good.

drgrice1 commented:
This looks good.  I am re-targeting for WeBWorK-2.16.

drgrice1 commented:
Nevermind.  Looks like that has been done.

Alex-Jordan commented on lib/WeBWorK/Authen.pm:
```diff
@@ -916,27 +882,37 @@ sub sendCookie {
+		-name     => "WeBWorKCourseAuthen.$courseID",
+		-value    => "$userID\t$key\t$timestamp",
+		-path     => $ce->{webworkURLRoot},
+		-samesite => $ce->{CookieSameSite};
```
This has a semicolon where a comma should be. I am getting errors restarting apache now.

Alex-Jordan replied:
(and same on next line)

pstaabp replied:
Trying to think of the best way to handle this.  Should I rebase this?  If I can remember how to?

taniwallach replied:
Ouch - that was a late change made late at night, but I did not see if trigger an error. Sorry.

Alex-Jordan commented on lib/WeBWorK/Authen.pm:
```diff
@@ -945,21 +921,22 @@ sub killCookie {
+		-expires   => "-1d", # 1 day ago
+		-path      => $ce->{webworkURLRoot},
+		-samesite  => $ce->{CookieSameSite};
+		-secure    => $ce->{CookieSecure};    # Warning: use 1 only if using https
```
Same here, with semicolons and commas.

pstaabp replied:
Alex, I just made a direct commit from the website.  Do a git pull  on the webwork2.16 branch and should be set.

drgrice1 commented:
Something is wrong here.  This pull request did not merge correctly.

pstaabp commented:
There were errors, but try a `git pull` on this branch and restart.

pstaabp commented:
I mean on `WeBWorK-2.16` branch.

drgrice1 commented:
@pstaab:  This is still not working.  No one do anything more.  I am working to fix it.  Don't merge anything.

pstaabp commented:
Sorry about that.  Not sure what I did.   I'm not having any troubles with the 2.16 branch.

drgrice1 commented:
I don't think it is what you did.  I think it is how @taniwallach redid the branch.

drgrice1 commented:
I think the only way to fix this is to reset the branch back to before this was merged.  The only thing after this was my TEXINPUTS change to the apache2 config file.   I will have to resubmit that, but @taniwallach is going to have to rework this pull request.  Do not merge it until I check it over.  I should have tested the merge before.

pstaabp commented:
What issue are you seeing?

drgrice1 commented:
It is done.  You will most likely need to run `git reset --hard origin WeBWorK-2.16` (and possible the same for develop) after running `git pull` (or better `git fetch`).

@pstaab: Almost none of this pull request actually went in.

@taniwallach:  I can help you fix this pull request.

Alex-Jordan commented:
What I did just now.

1. checkout develop
2. fetch origin (github.com/openwebwork/webwork2)
3. pull develop
4. delete local WeBWorK-2.16 branch
5. checkout new WeBWorK-2.16 from origin/WeBWorK-2.16

Should that leave me in good shape?

drgrice1 commented:
That will work.

pstaabp commented:
That's roughly what I did as well.

drgrice1 commented:
Okay.  This pull request is in now.  I had to rebuild the pull request into a new commit.

pstaabp commented:
@drgrice1 Are you saying this is now merged correctly in WeBWorK-2.16?

taniwallach commented:
I apologize for whatever made the mess, I had cherry-picked the commits from the old branch onto an update copy of develop (which was at bf11e63ee6bbe8ea9ee64e6c760f173e9c5ec8bb). That certainly worked on my branch, but apparently due to the merge+revert of the original branch - merging the new branch did not work.

Alex-Jordan commented:
No problem. But is someone able to assess if this will be resolved (as in, the intended commits get merged) within the hour or so? I could wait to being upgrading my production server, but not much longer than that. Is it helpful that this make it in, for field testing?

taniwallach commented:
> No problem. But is someone able to assess if this will be resolved (as in, the intended commits get merged) within the hour or so? I could wait to being upgrading my production server, but not much longer than that. Is it helpful that this make it in, for field testing?

@Alex-Jordan - I just checked out a clean copy of the WeBWorK-2.16 branch (after a git fetch) and compared using `diff -r` to the code in the branch this PR was supposed to be sourced from.  The code which this PR was supposed to modify is all done using @drgrice1's new 0db5cd6297b13c49684563c6dfe98755edb4cc73, and the other changes all seem related to several recently merged PRs, so it looks to me like everything in the WeBWorK-2.16 branch should be good now.

@drgrice1 Thanks for cleaning up my mess. Sorry. Do you know what caused the problem, so I can avoid it happening again.

pstaabp commented:
@taniwallach I think the cherry-picking can be dangerous.  In hindsight, you probably should have stayed with the original PR.   If a PR is going into a branch like `develop`, you can merge `develop` into that branch (especially if it has been a while since since branching from the original).

drgrice1 commented:
@taniwallach:  I am not entirely certain.  I just know that none of the config file changes actually appeared in the code after the merge was done.  The cherry-picking should have worked.  I have seen things like this before.  That is where GitHub shows things in the GUI in the browser, but what actually happens is different.

It is all fixed now, so no harm done.

taniwallach commented:
I don't think the original PR branch could work once the original PR was reverted. The commits were the same ones reverted - so apparently cannot be "reapplied" as is from the same "feature branch". When I quickly tried to make https://github.com/openwebwork/webwork2/pull/1253 and target master by accident and tried to retarget to develop it just made a huge mess. Apparently this is well-known to the experts, but confused plenty of people:

  - https://stackoverflow.com/questions/27852143/how-to-pr-and-merge-again-after-reverting-pr-using-github-revert-button
  - https://medium.com/@shanikae/create-pull-request-from-a-reverted-git-branch-27219cadf9b5
  - https://itnext.io/git-revert-the-revert-88b1e66d71d4

I suspect that the cherry picking from reverted commits is somehow a problem, but I am not certain.

If I am pulling out the correct lines from my `.bash_history` file, what I did was:
```
git checkout develop
git checkout -b cookie-refactor-same-site-v2
git checkout cookie-refactor-same-site
git log
git checkout cookie-refactor-same-site-v2
git cherry-pick edf75ca0d44c29de100f527ce3185a0b5bd03089..78b819dfc918ed7cb3ec3772e5b121927771c737
git status
vim conf/localOverrides.conf.dist
git add conf/localOverrides.conf.dist
git cherry-pick --continue
vim conf/defaults.config
git add conf/defaults.config
git cherry-pick --continue
vim lib/WeBWorK/Cookie.pm lib/WeBWorK/Authen.pm
git push --set-upstream tani cookie-refactor-same-site-v2
```
and then additional changes and commits were made.

---

We certainly have observed that comparing master and develop in GitHub shows something quite different from what `diff -r` would show.

---

I think we may find it safer to make small reverting changes directly using a new PR instead of reverting a merged PR for a small issue. The "bug" in the original PR (absolutely my fault) could have been corrected without reverting the entire PR.

Sorry again for all the mess.
